mars/dataframe/indexing/align.py

```diff
@@ -0,0 +1,554 @@
+            "dtypes": dtypes,
+            "shape": tuple(l_shape),
+            "columns_value": col_val,
+        }
+        if self.broadcast_axis == 1:
+            r_kws = {
+                "index_value": r_idx_val,
+                "dtypes": dtypes,
+                "shape": tuple(l_shape),
+                "columns_value": col_val,
+            }
+        else:
+            r_kws = {
+                "index_value": r_idx_val,
+                "shape": (r_size,),
+                "dtype": rhs.dtype,
+            }
+        return self.new_tileables([lhs, rhs], kws=[l_kws, r_kws])
+
+    def _call_series_series(self, lhs: TileableType, rhs: TileableType):
+        idx = self._merge_index(lhs.index_value, rhs.index_value, how=self.join)
+        kws = [
+            {"index_value": idx, "shape": (np.nan,), "dtype": lhs.dtype},
+            {"index_value": idx, "shape": (np.nan,), "dtype": rhs.dtype},
+        ]
+        return self.new_tileables([lhs, rhs], kws=kws)
+
+    @staticmethod
+    def _merge_index(
+        left_index_value: IndexValue, right_index_value: IndexValue, how: str = "outer"
+    ):
+        left_pd = left_index_value.to_pandas()
+        right_pd = right_index_value.to_pandas()
+
+        if not left_index_value.has_value() or not right_index_value.has_value():
+            left_pd = left_pd[:0]
+            right_pd = right_pd[:0]
+            store_data = False
+        else:
+            store_data = True
+
+        joined = left_pd.join(right_pd, how=how)
+        if store_data:
+            return parse_index(joined, store_data=store_data)
+        else:
+            return parse_index(
+                joined,
+                {left_index_value.key, right_index_value.key},
+                store_data=store_data,
+            )
+
+    @classmethod
+    def _select_nsplits(
+        cls, op: "DataFrameAlign", tileable: TileableType, val_to_replace: list
+    ):
+        if op.axis is None:
+            return val_to_replace[: tileable.ndim]
+        else:
+            attr_val = tileable.nsplits
+            axis = op.axis % tileable.ndim
+            return [
+                tuple(val_to_replace[op.axis]) if i == axis else attr_val[i]
+                for i in range(len(attr_val))
+            ]
+
+    @classmethod
+    def _build_tiled_kw(
+        cls, op: "DataFrameAlign", idx: int, chunks: list, nsplits: list
+    ):
+        in_tileable = op.inputs[idx]
+        out_tileable = op.outputs[idx]
+        kw = out_tileable.params.copy()
+        kw.update(
+            {
+                "chunks": chunks,
+                "nsplits": tuple(cls._select_nsplits(op, in_tileable, nsplits)),
+            }
+        )
+        return kw
+
+    @classmethod
+    def _check_align_needed(
+        cls, op: "DataFrameAlign", left_chunks: list, right_chunks: list
+    ):
+        lhs, rhs = op.lhs, op.rhs
+        if all(lc.key == rc.key for lc, rc in zip(lhs.chunks, left_chunks)) and all(
```

Comment:
`lc.index.key == rc.index.key` seems more reasonable, and if `lhs.index.key == rhs.index.key`, should we raise _NoNeedToAlign directly?

Reply:
Align methods in `mars.dataframe.align` already decides whether to align left or right chunks given indices. Thus we just check if chunks are changed after these calls.